ADD: EIP-5489 NFT Hyperlink Extension (#5489)
ethereum/EIPs: EIPS/eip-5489.md

---
eip: 5489
title: NFT hyperlink extension

> description: hyperlink extension for EIP-721

Author: The meaning can be extended by usages. For example, NFTs can be taken as billboards.

> description: hyperlink extension for EIP-721

Author: I've rephrase this description.

author: IronMan_CH (@coderfengyun)
discussions-to: https://ethereum-magicians.org/t/eip-5489-nft-hyperlink-extension/10431
status: Draft
type: Standards Track
category: ERC
created: 2022-08-16
requires: 165, 721
---

## Abstract

> This EIP proposes a new extension for NFTs (non-fungible token, aka [EIP-721](./eip-721.md)): nft-hyperlink-extention (hNFT), embedding NFTs with hyperlinks, referred to as “hNFTs”. As owners of hNFTs, users may authorize a URL slot to a specific address which can be either an externally-owned account (EOA) or a contract address and hNFT owners are entitled to revoke that authorization at any time. The address which has slot authorization can manage the URI of that slot.

Author: In my original thought, It's URL slot.

> This EIP proposes a new extension for NFTs (non-fungible token, aka [EIP-721](./eip-721.md)): nft-hyperlink-extention (hNFT), embedding NFTs with hyperlinks, referred to as “hNFTs”. As owners of hNFTs, users may authorize a URL slot to a specific address which can be either an externally-owned account (EOA) or a contract address and hNFT owners are entitled to revoke that authorization at any time. The address which has slot authorization can manage the URI of that slot.

Author: I've add usage in advertisement industry.

## Motivation
NFTs have entered mainstream consciousness in a big way as NFT trading volume spiked 21,000% to more than $17 billion in 2021. Despite NFTs’ monumental rise as blockchain-based digital assets, we are just starting to scratch the surface of what is possible with NFTs.

> The emergence of NFTs brings more possibilities for creators and Web3 projects to build up decentralized communities, which also shifts the previous company-consumer relationship to more of a partnership. Each NFT holder is entitled to trade, promote, or create based on their NFTs to make profits. 

Author: We rephrase this paragraph.

> The emergence of NFTs brings more possibilities for creators and Web3 projects to build up decentralized communities, which also shifts the previous company-consumer relationship to more of a partnership. Each NFT holder is entitled to trade, promote, or create based on their NFTs to make profits. 

Author: @SamWilsn I've rephrase motivation part.

However, there hasn’t been an effective approach to providing avenues of continued monetization for community holders. Going through all the NFT-related EIPs, the solutions to "making NFTs profitable" are limited to transfers ([EIP-721](./eip-721.md)), rental and mortgage ([EIP-4907](./eip-4907.md)).

> Inspired by hyperlink, the essence of World Wide Web, the proposed hNFT standard positions NFT as the medium for Web 3.0. The hNFT standard endows NFTs with a hyperlink extension so that it is possible to exercise the `fructus` right, meanwhile, keeps the `abusus`. With hyperlink extension, NFT owners can earn profit by authorizing any hyperlink slot to an address.

Reviewer: ```suggestion
```
Required: Not motivation.

Author: @Pandapip1 How about just keeping "With hyperlink extension, NFT owners can earn profit by authorizing any hyperlink slot to an address." this sentence ?

Reviewer: Again, that describes the solution, not the problem, and so would belong in the Abstract.

Author: I see, the core problems are:
1. no on-chain way to take advantage of NFT's influence without affecting ownership for now.
2. no standard on how to use in URI in NFT relative standard

Author: @Pandapip1 How about just list these two as the motivation?

Author: @Pandapip1 I've modified the motivation.

To establish standardization for NFT hyperlinks, it is important to share a unified standard so as to facilitate collaboration among all applications.

> Right to permission access to information—could be composed with “usus,” “abusus,” and “fructus” to create a nuanced constellation of access rights.

Author: This  paragraph is removed.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### NFT hyperlink extension

> This EIP extends EIP-721 with a `mapping<TokenId, mapping<SlotManagerAddress, uri>>`. The mapping means `TokenId` identified NFT has authorized `SlotManagerAddress` to manage the `uri` which is identified by `SlotManagerAddress`.

Author: Fixed

> NFT hyperlink extension extends EIP-721 with a `mapping<TokenId, mapping<SlotManagerAddress, uri>>`. The mapping means `TokenId` identified NFT has authorized `SlotManagerAddress` to manage the `uri` which is identified by `SlotManagerAddress`.

Reviewer: ```suggestion
This EIP extends EIP-721 with a `mapping<TokenId, mapping<SlotManagerAddress, uri>>`. The mapping means `TokenId` identified NFT has authorized `SlotManagerAddress` to manage the `uri` which is identified by `SlotManagerAddress`.
```
Required: Use "this EIP".

Besides, extension also provide the supporting facilities, such as
1. authorizeSlot/revokeSlot call pair used to manage slot authorization, and SlotAuthorizationCreate/Revoked event pair used to announce authorization management happens.

> 2. setSlotUri/getSlotUri call pair used to manage/view uri on slot, and SlotUriUpdated used to announce uri management happens.

Author: Fixed.

### New URI format

> This extension tries to add a new URI format, which is prefixed with "hnft://". This URI format enables redirect or connect activity between hNFTs, it enables great imagination.

Reviewer: ```suggestion
This EIP adds a new URI scheme, "hnft://". This URI scheme enables redirecting or connecting activity between hNFTs.
```
Required: Fix grammar.
Required: Use "this EIP".
Required: Use correct terminology.
Recommended: Rephrasing.

### Interface

#### `IERC5489`
```solidity
// SPDX-License-Identifier: CC0-1.0
pragma solidity ^0.8.0;

interface IERC5489 {
    /**
     * @dev this event emits when the slot on `tokenId` is authorzized to `slotManagerAddr`
     */
    event SlotAuthorizationCreated(uint256 tokenId, address slotManagerAddr);

    /**
     * @dev this event emits when the authorization on slot `slotManagerAddr` of token `tokenId` is revoked.
     * So, the corresponding DApp can handle this to stop on-going incentives or rights
     */
    event SlotAuthorizationRevoked(uint256 tokenId, address slotManagerAddr);

    /**
     * @dev this event emits when the uri on slot `slotManagerAddr` of token `tokenId` has been updated to `uri`.
     */

>     event SlotUriUpdated(uint256 tokenId, address slotManagerAddr, string uri);

Author: Will fix it.

>     event SlotUriUpdated(uint256 tokenId, address slotManagerAddr, string uri);

Author: I've fixed this.

     * This method will emit SlotAuthorizationCreated event
     */
    function authorizeSlotTo(uint256 tokenId, address slotManagerAddr) external;

    /**
     * @dev
     * Revoke the authorization of the slot indicated by `slotManagerAddr` on token `tokenId`
     * This method will emit SlotAuthorizationRevoked event
     */
    function revokeAuthorization(uint256 tokenId, address slotManagerAddr) external;

    /**
     * @dev
     * Revoke all authorizations of slot on token `tokenId`
     * This method will emit SlotAuthorizationRevoked event for each slot
     */
    function revokeAllAuthorizations(uint256 tokenId) external;

    /**
     * @dev
     * Set uri for a slot on a token, which is indicated by `tokenId` and `slotManagerAddr`
     * Only the address with authorization through {authorizeSlotTo} can manipulate this slot.
     * This method will emit SlotUriUpdated event
     */
    function setSlotUri(
        uint256 tokenId,
        string calldata newUri
    ) external;

    /**
     * @dev
     * returns the latest uri of an slot on a token, which is indicated by `tokenId`, `slotManagerAddr`
     */

>     function getSlotUri(uint256 tokenId, address slotManagerAddr)

Author: Do you mean [eip-3525](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3525.md) ?

        external
        view
        returns (string memory);
}
```

The `authorizeSlotTo(uint256 tokenId, address slotManagerAddr)` function MAY be implemented as public or external.

The `revokeAuthorization(uint256 tokenId, address slotManagerAddr)` function MAY be implemented as public or external.

The `revokeAllAuthorizations(uint256 tokenId)` function MAY be implemented as public or external.

The `setSlotUri(uint256 tokenId, string calldata newUri)` function MAY be implemented as public or external.

The `getSlotUri(uint256 tokenId, address slotManagerAddr)` function MAY be implemented as pure or view.

The `SlotAuthorizationCreated` event MUST be emitted when a slot is authorized to an address.

The `SlotAuthorizationRevoked` event MUST be emitted when a slot authorization is revoked.

The `SlotUriUpdated` event MUSt be emitted when a slot's URI is changed.

The `supportInterface` method MUST return true when called with `0x8f65987b`.

### Authentication
The `authorizeSlotTo`, `revokeAuthorization`, and `revokeAllAuthorizations` functions are authenticated if and only if the message sender is the owner of the token.

## Rationale

### Extends NFT with hyperlinks

> We use URI to represent value of slot to ensure it's flexible enough to deal with different use cases. E.g. one can define a custom uri format to indicate a brand new protocol.

Reviewer: ```suggestion
URIs are used to represent the value of slots to ensure enough flexibility to deal with different use cases.
```
Required: Fix grammar.
Recommended: Rephrasing.

### Authorize slot to address

> We use Address to represent key of slot to ensure it's flexible enough to deal with different use cases. E.g. one can manage slot's URI with an EOA, or manage slot's URI with a contract in which can do more detailed limitation or agreement.

Reviewer: ```suggestion
We use addresses to represent the key of slots to ensure enough flexibility to deal with all use cases.
```
Required: Fix grammar.
Recommended: Rephrasing.
